### PR TITLE
OpenTelemetry Integration

### LINKED ISSUE #72: Expose open telemetry metrics
### Proposed change

Follow the same pattern(s) as the NATS server (where/if appropriate) to allow Nex nodes to emit (or be scraped for) metrics. The ability to _export_ should be considered the primary goal (OTLP) while supporting _scraping_ should only be done if we get that for free or "for cheap" when we're setting up the metrics and exporting. We shouldn't use any specific vendor libraries like Jaeger or Prometheus, everything should be otel/OTLP standards.

Some gauges that I think we'll want ("deployed" includes all workload types, incl. functions):

- [x] deployed workload count
- [x] deployed workload count in namespace
- [x] deployed bytes count (size of total deployed workloads)
- [x] deployed bytes count in namespace
- [x] allocated memory total (based on firecracker machine config)
- [x] allocated memory in namespace
- [x] allocated vcpu count (based on fc vm config)
- [x] allocated vcpu count in namespace

Counters:

- [x] function trigger count total
- [x] function trigger count for workload
- [x] function trigger count for namespace
- [ ] function failure count total
- [ ] function failure count for workload
- [ ] function failure count for namespace
- [x] function execution elapsed time
- [x] function execution elapsed time for namespace
- [x] total function execution time across all funcs

### Use case

Operators managing a fleet of Nex nodes will need this feature

### Contribution

_No response_

### PR DESCRIPTION
**Pull request by jordan-rash**
Closes #72 

This will probably be around a week or so while we identify what metrics to expose.  This inital commit adds a metrics flag, and a "workload count" gauge.  This is a working PR, so please identify any helpful metrics in the comments.

### CODE REVIEW

**Comment by autodidaddict** [comment]
We want to make sure we don't have any prometheus dependencies. We should stick to using exporters and using only the abstractions available in otel, https://opentelemetry.io/docs/languages/go/

**Comment by jordan-rash** [comment]
> We want to make sure we don't have any prometheus dependencies. We should stick to using exporters and using only the abstractions available in otel, https://opentelemetry.io/docs/languages/go/

Alt Text

File: internal/node/machine_mgr.go [hunk 1]
```diff
@@ -131,17 +133,11 @@ func (m *MachineManager) Start() error {
...
-func (m *MachineManager) DeployWorkload(vm *runningFirecracker, workloadName, namespace string, request controlapi.RunRequest) error {
-	// TODO: make the bytes and hash/digest available to the agent
-	req := agentapi.DeployRequest{
-		WorkloadName:    &workloadName,
-		Hash:            nil, // FIXME
-		TotalBytes:      nil, // FIXME
-		TriggerSubjects: request.TriggerSubjects,
-		WorkloadType:    request.WorkloadType, // FIXME-- audit all types for string -> *string, and validate...
-		Environment:     request.WorkloadEnvironment,
+func (m *MachineManager) DeployWorkload(vm *runningFirecracker, runRequest controlapi.RunRequest, deployRequest agentapi.DeployRequest) error {
```
Comment (kthomas) [inline internal/node/machine_mgr.go:136]: It seems weird that we pass two different requests for `Deploy`. It was already weird that we were constructing a `DeployRequest` in this method, using mostly identically-named fields from the given `RunRequest`.

File: internal/node/controlapi.go [hunk 2]
```diff
@@ -217,10 +218,22 @@ func handleRun(api *ApiListener) func(m *nats.Msg) {
...
+			WorkloadName:    &workloadName,
+			Hash:            *workloadHash,
+			TotalBytes:      int64(numBytes),
+			TriggerSubjects: request.TriggerSubjects,
+			WorkloadType:    request.WorkloadType, // FIXME-- audit all types for string -> *string, and validate...
+			Environment:     request.WorkloadEnvironment,
+			Namespace:       &namespace,
+		}
+
+		err = api.mgr.DeployWorkload(runningVm, request, req)
```
Comment (kthomas) [inline internal/node/controlapi.go:236]: I don't love `request` and `req`. Very hard to read. We should be able to consolidate these right?

File: internal/node/machine_mgr.go [hunk 3]
```diff
@@ -242,6 +248,15 @@ func (m *MachineManager) Stop() error {
 	// Now empty the leftovers in the pool
 	for vm := range m.warmVms {
 		vm.shutDown(m.log)
+		// TODO: confirm this needs to be here
```
Comment (kthomas) [inline internal/node/machine_mgr.go:251]: We can probably DRY this up somehow.

File: internal/node/machine_mgr.go [hunk 4]
```diff
@@ -269,6 +284,15 @@ func (m *MachineManager) StopMachine(vmId string) error {
 	vm.shutDown(m.log)
 	delete(m.allVms, vmId)
 
+	workloadCounter.Add(m.rootContext, -1)
```
Comment (kthomas) [inline internal/node/machine_mgr.go:287]: We can probably DRY this up somehow.

File: internal/node/running_vm.go [hunk 5]
```diff
@@ -33,6 +33,7 @@ type runningFirecracker struct {
 	machineStarted        time.Time
 	workloadSpecification controlapi.RunRequest
 	namespace             string
+	deployedWorkload      agentapi.DeployRequest
```
Comment (kthomas) [inline internal/node/running_vm.go:36]: If we can consider how to consolidate the requests as mentioned above that would be fantastic. Here we refer to the `DeployRequest` as a `Workload` itself.

File: nex/devrunner.go [hunk 6]
```diff
@@ -52,7 +52,7 @@ func RunDevWorkload(ctx context.Context, logger *slog.Logger) error {
 	}
 	// developer mode can have a smaller discovery timeout, since we're assuming there's a NEX
 	// node "nearby"
-	nodeClient := controlapi.NewApiClientWithNamespace(nc, 750*time.Millisecond, "default", logger)
+	nodeClient := controlapi.NewApiClientWithNamespace(nc, 750*time.Millisecond, Opts.Namespace, logger)
```
Comment (kthomas) [inline nex/devrunner.go:55]: 👍🏻

File: nex/node_linux_amd64.go [hunk 7]
```diff
@@ -12,11 +22,45 @@ func setConditionalCommands() {
...
 
 	node_up.Flag("config", "configuration file for the node").Default("./config.json").StringVar(&NodeOpts.ConfigFilepath)
+	node_up.Flag("metrics", "enable open telemetry metrics endpoint").Default("false").UnNegatableBoolVar(&NodeOpts.OtelMetrics)
+	node_up.Flag("otel_metrics_exporter", "OTel exporter for metrics").Default("stdout").EnumVar(&NodeOpts.OtelMetricsExporter, "stdout", "prometheus")
 	node_preflight.Flag("force", "installs missing dependencies without prompt").Default("false").BoolVar(&NodeOpts.ForceDepInstall)
 	node_preflight.Flag("config", "configuration file for the node").Default("./config.json").StringVar(&NodeOpts.ConfigFilepath)
 }
 
 func RunNodeUp(ctx context.Context, logger *slog.Logger) error {
+	if NodeOpts.OtelMetrics {
```
Comment (kthomas) [inline nex/node_linux_amd64.go:32]: This code will live nicely inside the node initialization code once we are able to merge the `feat/node` branch. I can move it in there after we get this version of it merged 👍🏻

File: nex/node_linux_amd64.go [hunk 8]
```diff
@@ -31,3 +75,27 @@ func RunNodePreflight(ctx context.Context, logger *slog.Logger) error {
...
 	return nexnode.CmdPreflight(Opts, NodeOpts, ctx, cancel, logger)
 }
+
+func createMetricsReader(log *slog.Logger) (metric.Reader, error) {
+	switch NodeOpts.OtelMetricsExporter {
+	case "prometheus":
+		go func() {
+			log.Info("serving metrics at localhost:8085/metrics")
+			http.Handle("/metrics", promhttp.Handler())
+			err := http.ListenAndServe(":8085", nil)
```
Comment (kthomas) [inline nex/node_linux_amd64.go:85]: We probably make this port configurable.

File: internal/node/machine_mgr.go (hunk 1 above)
Comment (jordan-rash) [reply internal/node/machine_mgr.go:136]: I agree.  I was going to try and fix that in another PR so this one didn't start to balloon.  I need to create a issue for it

File: internal/node/controlapi.go (hunk 2 above)
Comment (jordan-rash) [reply internal/node/controlapi.go:236]: Agree.  See other comment about covering in different PR

File: internal/node/running_vm.go (hunk 5 above)
Comment (jordan-rash) [reply internal/node/running_vm.go:36]: Agree

File: nex/node_linux_amd64.go (hunk 7 above)
Comment (jordan-rash) [reply nex/node_linux_amd64.go:32]: Perfect

File: nex/node_linux_amd64.go (hunk 8 above)
Comment (jordan-rash) [reply nex/node_linux_amd64.go:85]: Yes.  I'll add that to this PR

File: internal/node/machine_mgr.go (hunk 1 above)
Comment (kthomas) [reply internal/node/machine_mgr.go:136]: On board with that.

**Review by kthomas — APPROVED** [review]

**Review by kthomas — APPROVED** [review]

**Pull request merged**